frame/mosaic/src/decay.rs

```diff
@@ -0,0 +1,99 @@
+use frame_support::pallet_prelude::*;
+use num_traits::{CheckedAdd, CheckedDiv, CheckedMul, CheckedSub, One, Saturating, Zero};
+
+pub trait Decayer<Balance, BlockNumber> {
+	fn checked_decay(
+		&self,
+		amount: Balance,
+		last_decay_block: BlockNumber,
+		current_block: BlockNumber,
+	) -> Option<Balance>;
+
+	fn full_recovery_period(&self, amount: Balance) -> Option<BlockNumber>;
+}
+
+/// Recommend type for storing the decay function of a penalty.
+#[derive(Decode, Encode, TypeInfo, Debug, PartialEq, Clone)]
+pub enum BudgetPenaltyDecayer<Balance, BlockNumber> {
+	/// Linear variant of the decay function, which decreases every block.
+	Linear(LinearDecay<Balance, BlockNumber>),
+}
+
+impl<Balance, BlockNumber> BudgetPenaltyDecayer<Balance, BlockNumber> {
+	#[allow(dead_code)]
+	pub fn linear(n: Balance) -> BudgetPenaltyDecayer<Balance, BlockNumber> {
+		BudgetPenaltyDecayer::Linear(LinearDecay { factor: n, _marker: PhantomData })
+	}
+}
+
+impl<Balance, BlockNumber> Decayer<Balance, BlockNumber>
+	for BudgetPenaltyDecayer<Balance, BlockNumber>
+where
+	BlockNumber: CheckedSub + Saturating + Into<Balance> + TryFrom<Balance> + One + CheckedAdd,
+	Balance: CheckedMul + CheckedDiv + Saturating + Zero,
+{
+	fn checked_decay(
+		&self,
+		amount: Balance,
+		last: BlockNumber,
+		current: BlockNumber,
+	) -> Option<Balance> {
+		match self {
+			BudgetPenaltyDecayer::Linear(lin) => lin.checked_decay(amount, last, current),
+		}
+	}
+
+	fn full_recovery_period(&self, amount: Balance) -> Option<BlockNumber> {
+		match self {
+			BudgetPenaltyDecayer::Linear(lin) => lin.full_recovery_period(amount),
+		}
+	}
+}
+
+#[derive(Decode, Encode, TypeInfo, Default, Debug, PartialEq, Clone)]
+pub struct LinearDecay<Balance, BlockNumber> {
+	/// Factor by which we decay every block.
+	factor: Balance,
+	_marker: core::marker::PhantomData<BlockNumber>,
+}
+
+impl<Balance, BlockNumber> Decayer<Balance, BlockNumber> for LinearDecay<Balance, BlockNumber>
+where
+	BlockNumber: CheckedSub + Saturating + Into<Balance> + TryFrom<Balance> + One + CheckedAdd,
+	Balance: CheckedMul + CheckedDiv + Saturating + Zero,
+{
+	fn checked_decay(
+		&self,
+		amount: Balance,
+		last: BlockNumber,
+		current: BlockNumber,
+	) -> Option<Balance> {
+		let diff = current.saturating_sub(last);
```

Comment:
safe_sub , do not hide bug behind 0 value

Reply:
please document return value of this trait that it is remaining penalty

Reply:
so if you pass last which is more than current, it would be better to return error to prove error in program, not hid and making diff = 0 always:(

Reply:
CU-1waqf8d